Mathlib/CategoryTheory/Limits/Shapes/Diagonal.lean

```diff
@@ -0,0 +1,461 @@
+
+end Diagonal
+
+end pullback
+
+variable [HasPullbacks C]
+
+open pullback
+
+section
+
+variable {U V₁ V₂ : C} (f : X ⟶ Y) (i : U ⟶ Y)
+
+variable (i₁ : V₁ ⟶ pullback f i) (i₂ : V₂ ⟶ pullback f i)
+
+@[reassoc (attr := simp)]
+theorem pullback_diagonal_map_snd_fst_fst :
+    (pullback.snd :
+          pullback (diagonal f)
+              (map (i₁ ≫ snd) (i₂ ≫ snd) f f (i₁ ≫ fst) (i₂ ≫ fst) i (by simp [condition])
+                (by simp [condition])) ⟶
+            _) ≫
+        fst ≫ i₁ ≫ fst =
+      pullback.fst := by
+  conv_rhs => rw [← Category.comp_id pullback.fst]
+  rw [← diagonal_fst f, pullback.condition_assoc, pullback.lift_fst]
+#align category_theory.limits.pullback_diagonal_map_snd_fst_fst CategoryTheory.Limits.pullback_diagonal_map_snd_fst_fst
+
+@[reassoc (attr := simp)]
+theorem pullback_diagonal_map_snd_snd_fst :
+    (pullback.snd :
+          pullback (diagonal f)
+              (map (i₁ ≫ snd) (i₂ ≫ snd) f f (i₁ ≫ fst) (i₂ ≫ fst) i (by simp [condition])
+                (by simp [condition])) ⟶
+            _) ≫
+        snd ≫ i₂ ≫ fst =
+      pullback.fst := by
+  conv_rhs => rw [← Category.comp_id pullback.fst]
+  rw [← diagonal_snd f, pullback.condition_assoc, pullback.lift_snd]
+#align category_theory.limits.pullback_diagonal_map_snd_snd_fst CategoryTheory.Limits.pullback_diagonal_map_snd_snd_fst
+
+variable [HasPullback i₁ i₂]
+
+set_option maxHeartbeats 400000 in
+/-- This iso witnesses the fact that
+given `f : X ⟶ Y`, `i : U ⟶ Y`, and `i₁ : V₁ ⟶ X ×[Y] U`, `i₂ : V₂ ⟶ X ×[Y] U`, the diagram
+
+V₁ ×[X ×[Y] U] V₂ ⟶ V₁ ×[U] V₂
+        |                 |
+        |                 |
+        ↓                 ↓
+        X         ⟶  X ×[Y] X
+
+is a pullback square.
+Also see `pullback_fst_map_snd_isPullback`.
+-/
+def pullbackDiagonalMapIso :
+    pullback (diagonal f)
+        (map (i₁ ≫ snd) (i₂ ≫ snd) f f (i₁ ≫ fst) (i₂ ≫ fst) i
+          (by simp only [Category.assoc, condition])
+          (by simp only [Category.assoc, condition])) ≅
+      pullback i₁ i₂
+    where
+  hom :=
+    pullback.lift (pullback.snd ≫ pullback.fst) (pullback.snd ≫ pullback.snd)
+      (pullback.hom_ext
+        (by
+           simp only [Category.assoc, pullback_diagonal_map_snd_fst_fst,
+              pullback_diagonal_map_snd_snd_fst])
+        (by simp only [Category.assoc, pullback.condition, pullback.condition_assoc]))
+  inv :=
+    pullback.lift (pullback.fst ≫ i₁ ≫ pullback.fst)
+      (pullback.map _ _ _ _ (𝟙 _) (𝟙 _) pullback.snd (Category.id_comp _).symm
+        (Category.id_comp _).symm)
+      (pullback.hom_ext
+        (by simp only [Category.assoc, diagonal_fst, Category.comp_id, limit.lift_π,
+          PullbackCone.mk_pt, PullbackCone.mk_π_app, limit.lift_π_assoc, cospan_left])
+        (by simp only [Category.assoc, diagonal_snd, Category.comp_id, pullback.lift_snd,
+            pullback.lift_snd_assoc, pullback.condition_assoc]))
+  hom_inv_id := by
+    apply pullback.hom_ext
```

Comment:
Can these be fixed by setting priorities on the `@[ext]` attributes. In mathlib3 I think there were no priorities and the most recent thing was used first, but in mathlib4 we should be setting priorities to make this work. You can write `@[ext 1100]` to increase the priority, and the default is `1000`.